actors/verifreg/src/lib.rs

```diff
@@ -366,108 +366,104 @@ impl Actor {
-    /// method returning an error.
+    /// Claims are processed in groups by sector. A failed claim will cause the
+    /// others in its group to fail too, unless `all_or_nothing` is enabled, in which case
+    /// the method will abort.
+    /// Returns an indicator of success for each sector group, and the size of claimed space.
     pub fn claim_allocations(
         rt: &impl Runtime,
         params: ClaimAllocationsParams,
     ) -> Result<ClaimAllocationsReturn, ActorError> {
         rt.validate_immediate_caller_type(std::iter::once(&Type::Miner))?;
         let provider = rt.message().caller().id().unwrap();
-        if params.allocations.is_empty() {
+        if params.sectors.is_empty() {
             return Err(actor_error!(illegal_argument, "claim allocations called with no claims"));
         }
 
-        let mut total_datacap_claimed = DataCap::zero();
-        let mut sector_claims = Vec::new();
-        let mut ret_gen = BatchReturnGen::new(params.allocations.len());
+        let mut batch_gen = BatchReturnGen::new(params.sectors.len());
+        let mut sector_results: Vec<SectorAllocationClaim> = vec![];
+        let mut total_claimed_space = DataCap::zero();
+
         rt.transaction(|st: &mut State, rt| {
             let mut claims = st.load_claims(rt.store())?;
             let mut allocs = st.load_allocs(rt.store())?;
 
-            for claim_alloc in params.allocations {
-                let maybe_alloc = state::get_allocation(
-                    &mut allocs,
-                    claim_alloc.client,
-                    claim_alloc.allocation_id,
-                )?;
-                let alloc: &Allocation = match maybe_alloc {
-                    None => {
-                        ret_gen.add_fail(ExitCode::USR_NOT_FOUND);
-                        info!(
-                            "no allocation {} for client {}",
-                            claim_alloc.allocation_id, claim_alloc.client,
-                        );
-                        continue;
+            // Note: this doesn't prevent being called with the same sector number twice.
+            'sectors: for sector in params.sectors {
+                // Load and validate all allocations for the sector group before
+                // making any state changes.
+                // Errors cause the sector to be skipped, unless all-or-nothing is requested.
+                let mut sector_new_claims: Vec<(ClaimID, Claim)> = vec![];
+                for claim in sector.claims {
+                    let maybe_alloc =
+                        state::get_allocation(&mut allocs, claim.client, claim.allocation_id)?;
+                    if let Some(alloc) = maybe_alloc {
+                        if !can_claim_alloc(&claim, provider, alloc, rt.curr_epoch(), sector.expiry)
+                        {
+                            info!("failed to claim allocation {}", claim.allocation_id);
```

Comment:
What's the reasoning behind not logging the failed sector id?  If anything I would have added sector id logging to the other error logs.

Reply:
Added the sector number and expiry.